### PR TITLE
request checkup

### PR DESCRIPTION
**Pull request by sanjoggit**
Currently stale checkup is done for publisher requests only whose state is new, i will implement later for other role after this is approved.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -28,7 +28,7 @@
 
 import {Utils} from '@natlibfi/identifier-services-commons';
 import Agenda from 'agenda';
-import {createRequestJobs, createMelindaJobs} from './jobs';
+import {createRequestJobs, createMelindaJobs, createCheckupJobs} from './jobs';
```
Comment (natlibfi-arlehiko) [inline src/index.js]: I think `createCleanupJobs` is more clear (Since we're not only checking, but acting on those checks)

File: src/index.js [hunk 2]
```diff
@@ -51,7 +51,9 @@ import {
 	JOB_PUBLICATION_ISSN_REQUEST_STATE_ACCEPTED,
 	JOB_PUBLICATION_ISSN_REQUEST_STATE_REJECTED,
 	JOB_BIBLIOGRAPHIC_METADATA_PENDING,
-	JOB_BIBLIOGRAPHIC_METADATA_INPROGRESS
+	JOB_BIBLIOGRAPHIC_METADATA_INPROGRESS,
+	REQUEST_TTL,
+	JOB_PUBLISHER_REQUEST_STATE_NEW_CHECK
```
Comment (natlibfi-arlehiko) [inline src/index.js]: Maybe a better name would be `JOB_REQUEST_BG_PROCESSING_CLEANUP`

File: src/jobs/checkup.js [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+export default function (agenda) {
+	const logger = createLogger();
+
+	const client = createApiClient({
+		url: API_URL, username: API_USERNAME, password: API_PASSWORD,
+		userAgent: API_CLIENT_USER_AGENT
+	});
+
+	agenda.define(JOB_PUBLISHER_REQUEST_STATE_NEW_CHECK, {concurrency: 1}, async (job, done) => {
```
Comment (natlibfi-arlehiko) [inline src/jobs/checkup.js]: No need for inline function here. Just use the function declared below.

File: src/jobs/checkup.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+		await request(job, done, 'new', 'publishers');
+	});
+
+	// eslint-disable-next-line max-params
+	async function request(job, done, state, type, subtype) {
+		try {
+			await getRequests();
```
Comment (natlibfi-arlehiko) [inline src/jobs/checkup.js]: I'd go for a few functions here:
1. Query for all requests which have the backgroundProcessingState set to pending
2. Filter out request that are have been updated frequently (Use TTL)
3. Process those requests that are left (Set the background processing state back to `pending`).

So the outer function will look something like this:
```js
try {
  const requests = await getRequests();
  const filteredRequests = await filterRequests();
  logger.log('debug', `${filteredRequests.length} requests need to have their background processing state set to 'pending'`);
  await processRequests();
}
```

File: src/jobs/cleanup.js [hunk 5]
```diff
@@ -27,61 +27,67 @@
...
+	});
+	agenda.define(JOB_REQUEST_BG_PROCESSING_CLEANUP_PUBLISHERS, {concurrency: 1}, async (_, done) => {
+		await request(done, 'publishers');
+	});
+	agenda.define(JOB_REQUEST_BG_PROCESSING_CLEANUP_ISBN_ISMN, {concurrency: 1}, async (_, done) => {
+		await request(done, 'publications', 'isbn-ismn');
+	});
+	agenda.define(JOB_REQUEST_BG_PROCESSING_CLEANUP_ISSN, {concurrency: 1}, async (_, done) => {
+		await request(done, 'publications', 'issn');
 	});
```
Comment (natlibfi-arlehiko) [inline src/jobs/cleanup.js]: Make this more compact:
```js
agenda.define(JOB_REQUEST_BG_PROCESSING_CLEANUP_ISSN, {concurrency: 1}, async (_, done) => request(done, 'publications', 'issn');
```

File: src/jobs/cleanup.js [hunk 6]
```diff
@@ -27,61 +27,67 @@
...
 
-	// eslint-disable-next-line max-params
-	async function request(job, done, state, type, subtype) {
+	async function request(done, type, subtype) {
```
Comment (natlibfi-arlehiko) [inline src/jobs/cleanup.js]: See my previous comments on implementing this function.

File: src/jobs/cleanup.js (hunk 5 above)
Comment (sanjoggit) [reply src/jobs/cleanup.js]: D you mean all 4 agenda or the last two?

Comment (natlibfi-arlehiko) [reply src/jobs/cleanup.js]: All. They all use the same function.

**Comment by sanjoggit** [comment]
> Still uses the old implementation. Use this kind of structure:
> 
> ```js
>   const requests = await getRequests();
>   const filteredRequests = await filterRequests();
>   logger.log('debug', `${filteredRequests.length} requests need to have their background processing state set to 'pending'`);
>   await processRequests();
> ```
I have changed according to the above structure, Can you check again :smile:

**Pull request merged**